### PR TITLE
feat: cross matching

### PR DESCRIPTION
**Pull request by pawel-betdex**
Changes:
- trade pk is now based of order pk (to which it belongs) and a distinct-seed (represented by 16 bytes array)

### CODE REVIEW

**Review by eoin-betdex — COMMENTED** [review]
A few comments so far. Looking good though I think. It will be good to see the new instructions in action.

File: npm-admin-client/src/market_management_instructions.ts [hunk 1]
```diff
@@ -175,7 +175,7 @@ export async function buildMarketManagementInstruction(
         return response.body;
       }
       const instruction = await program.methods
-        .openMarket()
+        .openMarket(false)
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply npm-admin-client/src/market_management_instructions.ts:178]: decided to expose the flag so we can later test it easier

File: npm-client/src/market_liquidities.ts [hunk 2]
```diff
@@ -72,3 +74,37 @@ export async function getMarketLiquidities(
...
+
+  try {
+    const enableCrossMatchingFilter = new BooleanCriterion(8 + 32);
+    enableCrossMatchingFilter.setValue(true);
+
+    const accounts = await connection.getProgramAccounts(program.programId, {
+      dataSlice: { offset: 0, length: 0 }, // fetch without any data.
+      filters: toFilters("market_liquidities", enableCrossMatchingFilter),
+    });
+    const accountKeys = accounts.map((account) => account.pubkey);
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply npm-client/src/market_liquidities.ts]: for now added `getCrossMatchEnabledMarketLiquiditiesPks`, still need to figure out what changes will be needed for the crank

File: programs/monaco_protocol/src/context.rs [hunk 3]
```diff
@@ -472,13 +472,67 @@ pub struct AuthoriseOperator<'info> {
...
+}
+
 fn market_matching_pool_constraint(
     market_matching_queue: &Account<MarketMatchingQueue>,
     market_matching_pool: &Account<MarketMatchingPool>,
 ) -> bool {
     match market_matching_queue.matches.peek() {
         Some(order_match) => {
-            market_matching_pool.for_outcome != order_match.for_outcome
+            market_matching_pool.for_outcome == order_match.for_outcome
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply programs/monaco_protocol/src/context.rs:535]: I know the name of the function is generic but not sure how to rename it without making it unmanageably long. I guess the issue is why it goes from `!=` to `==`. The long term solution would be to move all of this - functions and context - to it's own file like `programs/monaco_protocol/src/instructions/matching/process_order_match_maker_handler.rs`.

File: programs/monaco_protocol/src/instructions/matching/on_order_creation.rs [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
 
 pub fn on_order_creation(
     market_liquidities: &mut MarketLiquidities,
     market_matching_queue: &mut MarketMatchingQueue,
     order_pk: &Pubkey,
     order: &mut Order,
-) -> Result<Vec<OrderMatch>> {
+) -> Result<Vec<(u64, f64)>> {
     let mut order_matches = Vec::with_capacity(MATCH_CAPACITY);
+    let mut order_sources = Vec::with_capacity(MATCH_CAPACITY * 2);
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply programs/monaco_protocol/src/instructions/matching/on_order_creation.rs:18]: After reviewing this I've added following comment to the code
```
    // assuming every match is a cross-match we need n-1 times capacity for an n-way market
    // we only support 2 and 3 way markets so 2 covers our needs
```

File: programs/monaco_protocol/src/instructions/matching/on_order_match.rs [hunk 5]
```diff
@@ -15,94 +15,124 @@ use crate::state::trade_account::Trade;
...
             );
             market.increment_unclosed_accounts_count()?;
 
             emit!(TradeEvent {
-                amount: stake,
-                price: taker_order.price,
+                amount: matched_stake,
+                price: matched_price,
                 market: *market_pk,
             });
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply programs/monaco_protocol/src/instructions/matching/on_order_match.rs:132]: Taker is probably one we want as there is always a single taker in every match, while there can be multiple makers for cross-match.

File: programs/monaco_protocol/src/instructions/matching/on_order_creation.rs [hunk 6]
```diff
@@ -1,20 +1,22 @@
...
+
     let order_outcome = order.market_outcome_index;
 
     // FOR order matches AGAINST liquidity
```
[the remark this answers is not in the document]
Comment (pawel-betdex) [reply programs/monaco_protocol/src/instructions/matching/on_order_creation.rs:36]: done

File: admin/update_market_status.ts [hunk 7]
```diff
@@ -26,7 +26,7 @@ export async function openMarket() {
   const marketPk = new PublicKey(marketID);
 
   const protocolProgram = await getProtocolProgram();
-  checkResponse(await openMarketClient(protocolProgram, marketPk));
+  checkResponse(await openMarketClient(protocolProgram, marketPk, false));
```
Comment (pawel-betdex) [inline admin/update_market_status.ts:29]: expose this one

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 8]
```diff
@@ -122,19 +133,154 @@ impl MarketLiquidities {
...
+                    )
+                })
+                .min()
+                .unwrap_or(0_u64);
+
+            self.set_liquidity_against(outcome, cross_price, cross_liquidity, sources);
+        }
+    }
+
+    pub fn set_liquidity_for(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:202]: is it used only internally?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 9]
```diff
@@ -122,19 +133,154 @@ impl MarketLiquidities {
...
+            &mut self.liquidities_for,
+            sorter,
+            outcome,
+            price,
+            liquidity,
+            sources.to_vec(),
+        )
+    }
+
+    pub fn set_liquidity_against(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:220]: is it used only internally?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 10]
```diff
@@ -159,10 +304,11 @@ impl MarketLiquidities {
         }
     }
 
-    fn sorter_for(
+    pub fn sorter_for(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:307]: is it used only internally?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 11]
```diff
@@ -177,14 +323,23 @@ impl MarketLiquidities {
...
+            } else if liquidity_source_prices < source_prices {
+                return Ordering::Less;
+            }
+
             Ordering::Equal
         }
     }
 
-    fn sorter_against(
+    pub fn sorter_against(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:338]: is it used only internally?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 12]
```diff
@@ -199,18 +354,24 @@ impl MarketLiquidities {
...
-                .stake_matched_total
-                .checked_add(stake_matched)
-                .ok_or(CoreError::MarketLiquiditiesUpdateError)?;
-        }
-        Ok(())
+    fn source_prices(sources: &[LiquiditySource]) -> Vec<f64> {
+        sources.iter().map(|source| source.price).collect()
+    }
+
+    fn source_outcomes_sum(sources: &[LiquiditySource]) -> u16 {
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:373]: maybe we can pass the outcome rather calculate it?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 13]
```diff
@@ -222,113 +383,116 @@ impl MarketLiquidities {
...
+
     pub liquidity: u64,
 }
 
 impl MarketOutcomePriceLiquidity {
-    pub const SIZE: usize = U16_SIZE + F64_SIZE + U64_SIZE;
+    pub const SIZE: usize = U16_SIZE // outcome
+        + F64_SIZE // price
+        + U64_SIZE // liquidity
+        + vec_size(U16_SIZE + F64_SIZE, 3); // sources: sized to work for 3 and 4 way markets
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:411]: wrong order

File: programs/monaco_protocol/src/instructions/order/cancel_order.rs [hunk 14]
```diff
@@ -55,13 +55,15 @@ pub fn cancel_order(ctx: Context<CancelOrder>) -> Result<()> {
...
                     order.expected_price,
+                    &[],
                     order.voided_stake,
                 )
                 .map_err(|_| CoreError::CancelOrderNotCancellable)?,
             false => market_liquidities
                 .remove_liquidity_against(
                     order.market_outcome_index,
                     order.expected_price,
+                    &[],
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/instructions/order/cancel_order.rs:66]: tylko puste zrodla podawane?

File: programs/monaco_protocol/src/state/market_liquidities.rs (hunk 12 above)
Comment (pawel-betdex) [reply programs/monaco_protocol/src/state/market_liquidities.rs:344]: could do that but then needs to validate anyway - which is the same compute

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 15]
```diff
@@ -122,9 +133,114 @@ impl MarketLiquidities {
...
+        liquidities: &mut Vec<MarketOutcomePriceLiquidity>,
+        search_function: impl FnMut(&MarketOutcomePriceLiquidity) -> Ordering,
+        outcome: u16,
+        price: f64,
+        liquidity: u64,
+        sources: Vec<LiquiditySource>,
+    ) {
+        match liquidities.binary_search_by(search_function) {
+            Ok(index) => {
+                liquidities[index].liquidity = liquidity;
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:226]: remove when 0?

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 16]
```diff
@@ -122,9 +119,114 @@ impl MarketLiquidities {
...
+            .collect::<Vec<f64>>();
+        if let Some(cross_price) = calculate_price_cross(&source_prices) {
+            let cross_liquidity = sources
+                .iter()
+                .map(|source_liquidity_key| {
+                    let source_liquidity = self.get_liquidity_against(
+                        source_liquidity_key.outcome,
+                        source_liquidity_key.price,
+                    );
+                    calculate_stake_cross(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:140]: compute

File: programs/monaco_protocol/src/state/market_liquidities.rs [hunk 17]
```diff
@@ -122,9 +119,114 @@ impl MarketLiquidities {
...
+            .collect::<Vec<f64>>();
+        if let Some(cross_price) = calculate_price_cross(&source_prices) {
+            let cross_liquidity = sources
+                .iter()
+                .map(|source_liquidity_key| {
+                    let source_liquidity = self.get_liquidity_for(
+                        source_liquidity_key.outcome,
+                        source_liquidity_key.price,
+                    );
+                    calculate_stake_cross(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_liquidities.rs:180]: compute

File: programs/monaco_protocol/src/instructions/matching/on_order_creation.rs [hunk 18]
```diff
@@ -1,147 +1,729 @@
...
+mod test_match_for_order {
+    use crate::instructions::matching::on_order_creation;
+    use crate::instructions::matching::on_order_creation::{liquidities, liquidities2, matches};
+    use crate::state::market_liquidities::{mock_market_liquidities, LiquiditySource};
+    use crate::state::market_matching_queue_account::{MarketMatchingQueue, MatchingQueue};
+    use crate::state::order_account::mock_order;
+    use solana_program::pubkey::Pubkey;
+
+    #[test]
+    fn straight_match() {
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/instructions/matching/on_order_creation.rs:300]: direct

File: programs/monaco_protocol/src/state/market_matching_queue_account.rs [hunk 19]
```diff
@@ -112,8 +112,7 @@ impl MatchingQueue {
 
 #[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, Debug, Default)]
 pub struct OrderMatch {
-    pub pk: Pubkey,
-    pub purchaser: Pubkey,
+    pub pk: Option<Pubkey>,
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/state/market_matching_queue_account.rs:115]: comment taker vs maker

File: programs/monaco_protocol/src/instructions/order_request/process_order_request.rs [hunk 20]
```diff
@@ -86,12 +86,12 @@ pub fn process_order_request<'info>(
 
     // calculate payment
     let mut total_refund = 0_u64;
-    for order_match in &order_matches {
+    for (matched_stake, matched_price) in order_matches {
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/instructions/order_request/process_order_request.rs:89]: perhaps not needed change

File: programs/monaco_protocol/src/context.rs [hunk 21]
```diff
@@ -472,13 +472,67 @@ pub struct AuthoriseOperator<'info> {
...
+    #[account(mut)]
+    pub crank_operator: Signer<'info>,
+
+    #[account(address = system_program::ID)]
+    pub system_program: Program<'info, System>,
+    #[account(address = anchor_spl::token::ID)]
+    pub token_program: Program<'info, Token>,
+}
+
 fn market_matching_pool_constraint(
```
Comment (pawel-betdex) [inline programs/monaco_protocol/src/context.rs:529]: add a comment explainig who uses it

File: programs/monaco_protocol/src/instructions/order_request/process_order_request.rs (hunk 20 above)
Comment (pawel-betdex) [reply programs/monaco_protocol/src/instructions/order_request/process_order_request.rs:89]: `order_matches` used to be `Vec<OrderMatch>` but now it's `Vec<(u64, f64)` so this unpack is necessary to keep the code readable

**Review by stuartbetdex — APPROVED** [review]

**Review by obi-ewan — APPROVED** [review]

**Pull request merged**